[NNC] Preserve strided output

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#48264 [NNC] Preserve strided output**
* #48263 Remove inferred from tensor type ctors
* #47814 [TensorExpr Fuser] Add support for nodes which have tensor constant inputs
* #47813 [NNC] Compute Tensor Output Properties in ininitialization
* #47812 [TensorExpr] Cache use of fallback in kernel invocation
* #47810 Dont use symbolic shapes check

Preserves the strided representation of NNC Tensor outputs by transforming them into the right layout at the end of the kernel.

Fix for https://github.com/pytorch/pytorch/issues/45604

Differential Revision: [D25286213](https://our.internmc.facebook.com/intern/diff/D25286213)

Commit:
Preserve strided output

[ghstack-poisoned]

Commit:
Update on "[NNC] Preserve strided output"


Preserves the strided representation of NNC Tensor outputs by transforming them into the right layout at the end of the kernel.

[ghstack-poisoned]

Commit:
Update on "[NNC] Preserve strided output"


Preserves the strided representation of NNC Tensor outputs by transforming them into the right layout at the end of the kernel.

Fix for https://github.com/pytorch/pytorch/issues/45604

[ghstack-poisoned]

Commit:
Update on "[NNC] Preserve strided output"


Preserves the strided representation of NNC Tensor outputs by transforming them into the right layout at the end of the kernel.

Fix for https://github.com/pytorch/pytorch/issues/45604

[ghstack-poisoned]

Commit:
Update on "[NNC] Preserve strided output"


Preserves the strided representation of NNC Tensor outputs by transforming them into the right layout at the end of the kernel.

Fix for https://github.com/pytorch/pytorch/issues/45604

[ghstack-poisoned]

ZolotukhinM commented on test/cpp/tensorexpr/test_kernel.cpp:
```diff
@@ -139,6 +139,9 @@ TEST(Kernel, _3) {
 }
 
 TEST(Kernel, _4) {
+  // disabled: doesn't do stride propagation, and isn't being used currently
```
Maybe use `DISABLED_` feature of gtest instead? Also, from the comment and the test name (my fault, sorry), it's not clear *what* doesn't do stride propagation.

ZolotukhinM commented on test/cpp/tensorexpr/test_kernel.cpp:
```diff
@@ -685,8 +723,8 @@ TEST(Kernel, Softmax2D) {
       auto output = stack[0].toTensor();
-      auto ref =
-          log_softmax ? a.log_softmax(softmax_dim) : a.softmax(softmax_dim);
+      // auto ref =
```
Leftover?

ZolotukhinM commented on test/test_tensorexpr.py:
```diff
@@ -1505,6 +1505,39 @@ def simple(a, b):
 
+    def test_strided_output_preserved(self):
+        def foo(a, b):
+            return a + b - a
```
Nit: what's the reason of having two arguments if we're only using one? If having two is critical then it might be a good idea to pass two different inputs in the call site too.

ZolotukhinM reviewed (approved):
Looks good! A couple of minor comments inline.

eellison commented on torch/csrc/jit/tensorexpr/kernel.cpp:
```diff
@@ -1872,6 +1872,76 @@ std::vector<int64_t> TensorExprKernel::getReductionAxes(
+    return tensor;
+  }
+  auto dims = dimsFromSizes(sizesForValue(v));
+  // We need to convert the output tensor so that its values are layed
```
eek some of the wording/typos here can be cleaned up

Commit:
Update on "[NNC] Preserve strided output"


Preserves the strided representation of NNC Tensor outputs by transforming them into the right layout at the end of the kernel.

Fix for https://github.com/pytorch/pytorch/issues/45604

[ghstack-poisoned]

Commit:
Update on "[NNC] Preserve strided output"


Preserves the strided representation of NNC Tensor outputs by transforming them into the right layout at the end of the kernel.

Fix for https://github.com/pytorch/pytorch/issues/45604

Differential Revision: [D25286213](https://our.internmc.facebook.com/intern/diff/D25286213)

[ghstack-poisoned]

Commit:
Update on "[NNC] Preserve strided output"


Preserves the strided representation of NNC Tensor outputs by transforming them into the right layout at the end of the kernel.

Fix for https://github.com/pytorch/pytorch/issues/45604

Differential Revision: [D25286213](https://our.internmc.facebook.com/intern/diff/D25286213)

[ghstack-poisoned]

Commit:
Update on "[NNC] Preserve strided output"


Preserves the strided representation of NNC Tensor outputs by transforming them into the right layout at the end of the kernel.

Fix for https://github.com/pytorch/pytorch/issues/45604

Differential Revision: [D25286213](https://our.internmc.facebook.com/intern/diff/D25286213)

[ghstack-poisoned]

Commit:
Update on "[NNC] Preserve strided output"


Preserves the strided representation of NNC Tensor outputs by transforming them into the right layout at the end of the kernel.

Fix for https://github.com/pytorch/pytorch/issues/45604

Differential Revision: [D25286213](https://our.internmc.facebook.com/intern/diff/D25286213)

[ghstack-poisoned]

Commit:
Update on "[NNC] Preserve strided output"


Preserves the strided representation of NNC Tensor outputs by transforming them into the right layout at the end of the kernel.

Fix for https://github.com/pytorch/pytorch/issues/45604

Differential Revision: [D25286213](https://our.internmc.facebook.com/intern/diff/D25286213)

[ghstack-poisoned]
